### ISSUE TITLE
HttpEntity.Default body doesn't complete

### ISSUE THREAD

**Issue by rakesh2412** [akka/akka-http#750]
Using Akka 2.4.7 and akka-http-experimental_2.11 . I would like to process http request by reading the request body and passing the entity  to underlaying akka runtime to further processing before sending an HttpResponse.

```
if (request.method() == HttpMethods.POST) {
	CompletionStage<ByteString> body =  
         request.entity().getDataBytes().runWith(Sink.head(),materializer);
	returnResponse = HttpResponse.create().withEntity(handleMessage(body.toCompletableFuture().
       get().decodeString("UTF-8").toString()));
}
```

This works well if the POST request has a small payload. 

When Request payload larger than 1.2 K , the body of an HttpEntity.Default does complete or end up with the below exception.

scala.concurrent.java8.FuturesConvertersImpl$CF@4aaf3b7c[Not completed]    // HttpEntity.Default does complete
-- OR
java.lang.IllegalStateException: Substream Source cannot be materialized more than once  // Exception

I also tried "akka.http.parsing.max-content-length = 8m" and withoutSizeLimit() , but it is still not working.

Observation:
1) request payload < 1.2k   =  HttpEntity.Strict(application/json
2) request payload > 1.2k  = HttpEntity.Default(application/json,1708715

What could be the reason for a different entity type?  Is there any simple workaround I can implement on my side?

Here are the full requests 
// request payload < 1.2k   --- Good case

HttpEntity.Strict(application/json,{"command":"start","UserDetail":{  ........

//  request payload > 1.2k

> HttpEntity.Default(application/json,1708715,Source(SourceShape(StreamUtils$$anon$2.out), CompositeModule [0a052a3a]
>   Name: unnamed
>   Modules:
>     (unnamed) CompositeModule [095f6277]
>       Name: unnamed
>       Modules:
>         (SubSource%28EntitySource%29) GraphStage(EntitySource) [14e23e14]
>         (unnamed) [67a1e2a4] copy of GraphStage(Collect) [017513e6]
>         (limitable) [73cb8565] copy of CompositeModule [765c1796]
>           Name: limitable
>           Modules:
>             (unnamed) GraphStage(unknown-operation) [1e806f3b]
>           Downstreams: 
>           Upstreams: 
>           MatValue: Ignore
>       Downstreams: 
>         SubSource.out -> Collect.in
>         Collect.out -> unknown-operation.in
>       Upstreams: 
>         Collect.in -> SubSource.out
>         unknown-operation.in -> Collect.out
>       MatValue: Atomic(SubSource%28EntitySource%29[14e23e14])
>     (unnamed) [1d2dd471] copy of GraphStage(akka.http.impl.util.StreamUtils$$anon$2@36aa198e) [797c6f34]
>   Downstreams: 
>     SubSource.out -> Collect.in
>     Collect.out -> unknown-operation.in
>     unknown-operation.out -> StreamUtils$$anon$2.in
>   Upstreams: 
>     Collect.in -> SubSource.out
>     unknown-operation.in -> Collect.out
>     StreamUtils$$anon$2.in -> unknown-operation.out
>   MatValue: Atomic(akka.stream.impl.StreamLayout$CompositeModule[095f6277])))

**Comment by jrudolph** [comment]
Hi @rakesh2412,

can you update to the most recent version of akka-http, 10.0.1 and see if the error persists? Also see, the documentation about streaming entities at http://doc.akka.io/docs/akka-http/current/java/http/common/http-model.html#httpentity and http://doc.akka.io/docs/akka-http/current/java/http/implications-of-streaming-http-entity.html.

**Comment by jrudolph** [comment]
With

```java
request.entity().getDataBytes().runWith(Sink.head(),materializer)
```

you are also only reading the first chunk of the data. Is that what you want? What should happen with the data?

**Comment by rakesh2412** [comment]
Thanks a lot for the quick response!.

I have updated recent version of akka-http, 10.0.1 but still no luck.

All I need to consume the entire request body(json data) and passing the entity to underlaying akka runtime to further processing before sending an HttpResponse.

I also tried the following approaches but it does not work for me.

1) Akka HTTP toStrict(timeout, materializer) method to eagerly consume the entity and make it available in memory:

2) Framing the incoming chunks, parsing and then connecting the flow into another destination Sink, such as a File

```
final Function<ByteString, ByteString> transformEachLine = line -> line 
final int maximumFrameLength = 4096;

CompletionStage<IOResult> body = request.entity().getDataBytes()
	.via(Framing.delimiter(ByteString.fromString("\n"), maximumFrameLength, FramingTruncation.ALLOW))
	.map(transformEachLine::apply)
	.runWith(FileIO.toPath(new File("/tmp/response.txt").toPath()), materializer);
```
// Exception:  				    
```
Error in stage [akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage@2634d4c8]: requirement failed: Cannot push port (requestPrepOut) twice
java.lang.IllegalArgumentException: requirement failed: Cannot push port (requestPrepOut) twice
	at scala.Predef$.require(Predef.scala:224)
	at akka.stream.stage.GraphStageLogic.push(GraphStage.scala:436)
	at akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage$$anon$12$$anon$13.onPush(HttpServerBluePrint.scala:395)
	at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:590)
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:601)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:542)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:472)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:424)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:604)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:619)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:530)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jrudolph** [comment]
Ok, that may be the bug we have noted in various occasions. Search for "requestPrepOut" on the tracker. Could you include more information how to reproduce this? Are you using `curl` to issue the requests?

**Comment by rakesh2412** [comment]
I'm sorry for taking so long to write back. I tried to create a reproducer related with this issue.

I've created a test project that demonstrates the problem. You can run it and test it with curl.

[http-test-tool.zip](https://github.com/akka/akka-http/files/717485/http-test-tool.zip)

Although the actual code looks nothing like this (goes through several other APIs) that is why this code looks so weird and unnecessary.

I would like to process http request by reading the request body entirely and passing the entity to underlaying akka runtime to further processing before sending an HttpResponse.


Following curl command to send a request with an HttpEntity.Default request body. Request.json file being used in curl command is included in "src/main/resource"
folder. 

curl -v -T request.json  -X POST "http://localhost:1111/"

**Comment by jrudolph** [comment]
Then it's probably a duplicate of https://github.com/akka/akka-http/issues/396.

**Comment by jrudolph** [comment]
We will release a fix for #396 / #516 with 10.0.4. Can you try again once it is released?

I'm closing this for now as a duplicate of #396 / #516. Feel free to reopen if the issue persists in 10.0.4.
